File: BST.rs
use std::fmt::Display;
use std::fmt::Debug;

#[derive(Clone, Debug)]
struct TreeNode<T> {
    pub left: Option<Box<TreeNode<T>>>,
    pub right: Option<Box<TreeNode<T>>>,
    pub value: T,
}

#[derive(Clone, Debug)]
struct BinaryTree<T> {
    root: TreeNode<T>,
}

impl<T: Display> BinaryTree<T> {
    pub fn print_in_order(&self, root: &TreeNode<T>) {
        if let Some(left) = &root.left {
            self.print_in_order(left);
        }
        print!("{}, ", root.value);
        if let Some(right) = &root.right {
            self.print_in_order(right);
        }
    }
    pub fn print_pre_order(&self, root: &TreeNode<T>) {
        print!("{}, ", root.value);
        if let Some(left) = &root.left {
            self.print_in_order(left);
        }
        if let Some(right) = &root.right {
            self.print_in_order(right);
        }
    }
    pub fn print_post_order(&self, root: &TreeNode<T>) {
        if let Some(left) = &root.left {
            self.print_in_order(left);
        }
        if let Some(right) = &root.right {
            self.print_in_order(right);
        }
        print!("{}, ", root.value);
    }
    pub fn print_out_order(&self, root: &TreeNode<T>) {
        if let Some(right) = &root.right {
            self.print_in_order(right);
        }
        print!("{}, ", root.value);
        if let Some(left) = &root.left {
            self.print_in_order(left);
        }
    }
}
impl<T> TreeNode<T> {
    fn new(val: T) -> Self {
        return TreeNode{value: val, left: None, right: None};
    }
}
fn main() {
    let mut tree = BinaryTree{root: TreeNode::new(9)};
    tree.root.left = Some(Box::new(TreeNode::new(3)));
    tree.root.right = Some(Box::new(TreeNode::new(17)));
    tree.print_in_order(&tree.root);
    println!();
}

File: BST/BST.java
/**
 * @author Sean Hellum
 * <p>This is a BST library</p>*/
import java.util.List;
import java.util.Arrays;
public class BST {
    public static class BinaryTree {
        Node root;
        BinaryTree(double value) {
            root = new Node(value);
        }
        BinaryTree() {
            root = null;
        }
    }
    public static class Node {
        double value;
        Node left, right;
        public Node(double item) {
            value = item;
            left = null;
            right = null;
        }
    }
    /**
     * @param node The root node of the tree i.e. printInorder(tree.root);*/
    public static void printInorder(Node node) {
        if (node == null) {
            return;
        }
        printInorder(node.left);
        System.out.print(node.value + ",");
        printInorder(node.right);
    }
    public static void printPreorder(Node node) {
        if (node == null) {
            return;
        }
        System.out.print(node.value + ",");
        printPreorder(node.left);
        printPreorder(node.right);
    }
    public static void deleteTree(Node node) {
        node=null;
    }
    public static void printPostorder(Node node) {
        if (node == null) {
            return;
        }
        printPostorder(node.left);
        printPostorder(node.right);
        System.out.print(node.value + ",");
    }
    public static void printOutOrder(Node node) {
        if (node == null) {
            return;
        }
        printOutOrder(node.right);
        System.out.print(node.value + ",");
        printOutOrder(node.left);
    }
    public static void prettyPrintInOrder(Node node) {
        printInorder(node);
        System.out.println();
    }
    public static void prettyPrintPreOrder(Node node) {
        printPreorder(node);
        System.out.println();
    }
    public static void prettyPrintPostOrder(Node node) {
        printPostorder(node);
        System.out.println();
    }
    public static void prettyPrintOutOrder(Node node) {
        printOutOrder(node);
    }
    public static boolean findNode(double target,Node current) {
        if (target == current.value) {
            return true;
        } else if (target < current.value) {
            if(current.left!=null) {
                current = current.left;
            } else {
                return false;
            }
        } else if (target >= current.value) {
            if (current.right!=null) {
                current = current.right;
            } else {
                return false;
            }
        }
        return findNode(target, current);
    }
    public static Node insertNode(Node node, Node root) {
        if (root == null) {
           return node;
        } else if (node.value < root.value) {
            root.left = insertNode(node, root.left);
        } else {
            root.right = insertNode(node, root.right);
        }
        return root;
    }
    public static BinaryTree assembleTree(List<Double> list) {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(list.get(0));
        int index = 1;
        while(index<=list.size()-1) {
            insertNode(new Node(list.get(index)), tree.root);
            index++;
        }
        return tree;
    }
}